### PR TITLE
Schema  and Credential Definition validations in VDR

### PR DESCRIPTION
**Pull request by DenisRybas**
Added schema and Credential Definition validations to VDR (according to [the previous smart-contracts logic](https://github.com/DSRCorporation/indy-node/blob/096fb303448d644f75c33f6c7dec6a5b192e6774/indy-besu/smart_contracts/contracts/cl/SchemaValidator.sol))

### CODE REVIEW

File: vdr/src/contracts/cl/types/schema.rs [hunk 1]
```diff
@@ -17,6 +17,55 @@ pub struct Schema {
...
+        let actual_schema_id = SchemaId::build(&self.issuer_id, &self.name, &self.version);
+
+        if expected_schema_id != actual_schema_id {
+            return Err(VdrError::InvalidSchema(actual_schema_id.to_string()));
+        }
+
+        Ok(())
+    }
+
+    fn require_name(&self) -> VdrResult<()> {
```
Comment (Artemkaaas) [inline vdr/src/contracts/cl/types/schema.rs:31]: Just put all checks inside of `validate` function. Do not see a big sense in defining separate functions.

File: vdr/src/contracts/cl/types/schema.rs [hunk 2]
```diff
@@ -17,6 +17,55 @@ pub struct Schema {
     pub attr_names: Vec<String>,
 }
 
+impl Schema {
+    pub fn require_valid_id(&self, expected_schema_id: &SchemaId) -> VdrResult<()> {
+        let actual_schema_id = SchemaId::build(&self.issuer_id, &self.name, &self.version);
+
+        if expected_schema_id != actual_schema_id {
+            return Err(VdrError::InvalidSchema(actual_schema_id.to_string()));
```
Comment (Artemkaaas) [inline vdr/src/contracts/cl/types/schema.rs:25]: please add error description

File: vdr/src/contracts/cl/types/schema.rs [hunk 3]
```diff
@@ -17,6 +17,55 @@ pub struct Schema {
     pub attr_names: Vec<String>,
 }
 
+impl Schema {
+    pub fn require_valid_id(&self, expected_schema_id: &SchemaId) -> VdrResult<()> {
```
Comment (Artemkaaas) [inline vdr/src/contracts/cl/types/schema.rs:21]: let add method just returning`id` for schema and move comparison

**Review by Artemkaaas — CHANGES_REQUESTED** [review]

File: vdr/src/contracts/cl/types/schema.rs (hunk 1 above)
Comment (DenisRybas) [reply vdr/src/contracts/cl/types/schema.rs]: Moved checks

File: vdr/src/contracts/cl/types/schema.rs (hunk 2 above)
Comment (DenisRybas) [reply vdr/src/contracts/cl/types/schema.rs]: Added

File: vdr/src/contracts/cl/types/schema.rs (hunk 3 above)
Comment (DenisRybas) [reply vdr/src/contracts/cl/types/schema.rs]: Added

File: vdr/src/contracts/cl/schema_registry.rs [hunk 4]
```diff
@@ -298,6 +301,45 @@ pub mod test {
             };
             assert_eq!(expected_transaction, transaction);
         }
+
+        #[async_std::test]
```
Comment (Artemkaaas) [inline vdr/src/contracts/cl/schema_registry.rs:305]: Combine using rstest

File: vdr/src/contracts/cl/types/credential_definition.rs [hunk 5]
```diff
@@ -24,6 +24,45 @@ pub enum CredentialDefinitionTypes {
...
+    pub fn id(&self) -> CredentialDefinitionId {
+        CredentialDefinitionId::build(&self.issuer_id, &self.schema_id.to_string(), &self.tag)
+    }
+
+    pub fn matches_id(&self, expected_id: &CredentialDefinitionId) -> VdrResult<()> {
+        let actual_id = self.id();
+
+        if expected_id.to_string() != actual_id.to_string() {
+            return Err(VdrError::InvalidCredDef(format!(
+                "Id built from cred_def: {} != provided id: {}",
```
Comment (Artemkaaas) [inline vdr/src/contracts/cl/types/credential_definition.rs:37]: Credential Definition ID `{}` does not match to requested `{}`

File: vdr/src/contracts/cl/types/credential_definition.rs [hunk 6]
```diff
@@ -24,6 +24,45 @@ pub enum CredentialDefinitionTypes {
...
+                actual_id.to_string(),
+                expected_id.to_string()
+            )));
+        }
+
+        Ok(())
+    }
+
+    pub fn validate(&self) -> VdrResult<()> {
+        if self.cred_def_type != CredentialDefinitionTypes::CL {
```
Comment (Artemkaaas) [inline vdr/src/contracts/cl/types/credential_definition.rs:47]: Please rebase your branch on `DSRCorporation:feature/legacy-ids-mapping`. 
I partially integrated `indy-data-types` and some of the checks will go away.

File: vdr/src/contracts/cl/types/schema.rs [hunk 7]
```diff
@@ -17,6 +17,46 @@ pub struct Schema {
...
+    pub fn id(&self) -> SchemaId {
+        SchemaId::build(&self.issuer_id, &self.name, &self.version)
+    }
+
+    pub fn matches_id(&self, expected_id: &SchemaId) -> VdrResult<()> {
+        let actual_id = self.id();
+
+        if expected_id.to_string() != actual_id.to_string() {
+            return Err(VdrError::InvalidSchema(format!(
+                "Id built from schema: {} != provided id: {}",
```
Comment (Artemkaaas) [inline vdr/src/contracts/cl/types/schema.rs:30]: I think we can delete this method (and for cred def) as it's called from a single place. and move check there

**Review by Artemkaaas — CHANGES_REQUESTED** [review]

File: vdr/src/contracts/cl/schema_registry.rs (hunk 4 above)
Comment (DenisRybas) [reply vdr/src/contracts/cl/schema_registry.rs]: combined

File: vdr/src/contracts/cl/schema_registry.rs [hunk 8]
```diff
@@ -38,7 +38,7 @@ pub async fn build_create_schema_transaction(
     id: &SchemaId,
```
[the remark this answers is not in the document]
Comment (DenisRybas) [reply vdr/src/contracts/cl/schema_registry.rs]: Done

File: vdr/src/contracts/cl/types/credential_definition.rs (hunk 5 above)
Comment (DenisRybas) [reply vdr/src/contracts/cl/types/credential_definition.rs]: Changed description

File: vdr/src/contracts/cl/types/credential_definition.rs (hunk 6 above)
Comment (DenisRybas) [reply vdr/src/contracts/cl/types/credential_definition.rs]: Rebased. Removed cred_def_type check

File: vdr/src/contracts/cl/types/schema.rs (hunk 7 above)
Comment (DenisRybas) [reply vdr/src/contracts/cl/types/schema.rs]: Deleted

**Review by Artemkaaas — APPROVED** [review]

**Pull request merged**